pkg/external_repos/introspect.go

```diff
@@ -132,3 +147,33 @@ func httpClient(useCert bool) (http.Client, error) {
 		return http.Client{}, nil
 	}
 }
+
+// UpdateIntrospectionStatusMetadata updates introspection timestamps, error, and status on repo. Use after calling Introspect().
+func UpdateIntrospectionStatusMetadata(repo dao.Repository, repoDao dao.RepositoryDao, count int64, err error) error {
+	introspectTimeEnd := time.Now()
+	repo.LastIntrospectionTime = &introspectTimeEnd
+
+	if err != nil {
+		repo.LastIntrospectionError = pointy.String(err.Error())
+
+		switch repo.Status {
+		case "Pending":
+			repo.Status = "Invalid"
+		default:
+			repo.Status = "Unavailable"
+		}
+	} else {
+		if count != 0 {
+			repo.LastIntrospectionUpdateTime = &introspectTimeEnd
+		}
+		repo.LastIntrospectionSuccessTime = &introspectTimeEnd
+		repo.LastIntrospectionError = pointy.String("")
+		repo.Status = "Valid"
+	}
```

Comment:
I suggest to extract this block in a private function and add a unit test to cover it; suggested block:

```golang
// I am sure a better name for this private function, only to express the idea
func introspectTransition(input dao.Repository, count int64, err error, introspectTimeEnd *time.Time) dao.Repository {
	output := input

	output.LastIntrospectionTime = introspectTimeEnd

	if err == nil {
		if count != 0 {
			output.LastIntrospectionUpdateTime = introspectTimeEnd
		}
		output.LastIntrospectionSuccessTime = introspectTimeEnd
		output.LastIntrospectionError = pointy.String("")
		output.Status = StatusValid
		return output
	}

	output.LastIntrospectionError = pointy.String(err.Error())

	switch input.Status {
	case StatusValid:
		output.Status = StatusUnavailable
		break
	case StatusInvalid:
		output.Status = StatusUnavailable
		break
	case StatusPending:
		output.Status = StatusInvalid
		break
	case StatusUnavailable:
		// Do nothing, but express the status don't change
		break
	default:
		output.Status = StatusUnavailable
		break
	}
	return output
}
```

I know the cases can be simplified, but in this specific case IMO I would leave all the cases to express the state transition.

Reply:
Sorry, I think I wrote wrong the above; I think the transition `Invalid` > `Unavailable` is `Invalid` > `Invalid`.

Reply:
this is a good idea! changed it to this and added tests